app/src/main/java/org/wikipedia/page/linkpreview/LinkPreviewDialog.kt

```diff
@@ -106,31 +100,105 @@ class LinkPreviewDialog : ExtendedBottomSheetDialogFragment(), LinkPreviewErrorV
             }
         }
-        L10nUtil.setConditionalLayoutDirection(binding.root, pageTitle.wikiSite.languageCode)
-        loadContent()
-        funnel = LinkPreviewFunnel(WikipediaApp.instance, historyEntry.source)
+        L10nUtil.setConditionalLayoutDirection(binding.root, viewModel.pageTitle.wikiSite.languageCode)
+        renderViewStates()
+//        viewModel.loadContent()
+        funnel = LinkPreviewFunnel(WikipediaApp.instance, viewModel.historyEntry.source)
         funnel.logLinkClick()
         return binding.root
     }
 
+    private fun renderViewStates() {
+        lifecycleScope.launchWhenCreated {
+            viewModel.uiState.collect {
+                when (it) {
+                    is LinkPreviewViewState.Loading -> {
+                        binding.linkPreviewProgress.visibility = View.VISIBLE
+                    }
+                    is LinkPreviewViewState.Error -> {
+                        renderErrorState(it.throwable)
+                    }
+                    is LinkPreviewViewState.Content -> {
+                        renderContentState(it.data)
+                    }
+                    is LinkPreviewViewState.Gallery -> {
+                        renderGalleryState(it)
+                    }
+                    is LinkPreviewViewState.Completed -> {
+                        binding.linkPreviewProgress.visibility = View.GONE
+                    }
+                }
+            }
+        }
+    }
+
+    private fun renderGalleryState(it: LinkPreviewViewState.Gallery) {
+        binding.linkPreviewThumbnailGallery.setGalleryList(it.data)
+        binding.linkPreviewThumbnailGallery.listener = galleryViewListener
+        binding.linkPreviewProgress.visibility = View.GONE
+    }
+
+    private fun renderContentState(response: Response<PageSummary>) {
+        val summary = response.body()!!
+        funnel.setPageId(summary.pageId)
+        articleLinkPreviewInteractionEvent = ArticleLinkPreviewInteractionEvent(
+                viewModel.pageTitle.wikiSite.dbName(),
+                summary.pageId,
+                viewModel.historyEntry.source
+        )
+        articleLinkPreviewInteractionEvent?.logLinkClick()
+        revision = summary.revision
+
+        // Rebuild our PageTitle, since it may have been redirected or normalized.
+        val oldFragment = viewModel.pageTitle.fragment
+        viewModel.pageTitle = PageTitle(
+                summary.apiTitle, viewModel.pageTitle.wikiSite, summary.thumbnailUrl,
+                summary.description, summary.displayTitle
+        )
+
+        // check if our URL was redirected, which might include a URL fragment that leads
+        // to a specific section in the target article.
+        if (!response.raw().request.url.fragment.isNullOrEmpty()) {
+            viewModel.pageTitle.fragment = response.raw().request.url.fragment
+        } else if (!oldFragment.isNullOrEmpty()) {
+            viewModel.pageTitle.fragment = oldFragment
+        }
```

Comment:
This can all be done in the ViewModel, and then pass back the `PageSummary` instead of `Response<PageSummary>`.